.github/workflows/rpython-unit-tets.yml

```diff
@@ -0,0 +1,174 @@
+name: Rpython Unit Tests
+# === Implementation notes ===
+# Since the test suite is too large to run in a reasonable amount of time
+# we split this into tests by sub-folder within the rpython folder
+
+# Note: this action will only run if there are changes to one of:
+#  - rpython folder
+#  - .github folder
+#  - pytest.py
+# add more items to `paths` if we missed some!
+
+on:
+  push:
+    branches: [ "main", "py3.9", "py3.10" ]
+    paths:
+        - "rpython/**/*"
+        - "pytest.py"
+        - ".github/**/*"
+  pull_request:
+    branches: [ "main", "py3.9", "py3.10" ]
+    paths:
+      - "rpython/**/*"
+      - "pytest.py"
+      - ".github/**/*"
+  workflow_dispatch:
+
+# Limit tests to latest commit on branches other than main
+concurrency:
+  group: ${{ github.ref }}
+  cancel-in-progress: ${{ github.ref != 'refs/heads/main' }}
+
+
+permissions:
+  contents: read
+
+jobs:
+  test-misc:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-20.04]
+
+    steps:
+    - uses: actions/checkout@v4
+    - uses: actions/setup-python@v5
+      with:
+        python-version: "pypy2.7-7.3.10"
+    - run: |
+          echo "Install requirements"
+          sudo apt-get install -y libgc-dev
+          pypy2 -m pip install --user hypothesis==4.39.3
+    - name: Test Annotator
+      run: pypy2 pytest.py rpython/annotator
+    - name: Test Config
+      run: pypy2 pytest.py rpython/config
+    - name: Test Flowspace
+      run: pypy2 pytest.py rpython/flowspace
+    - name: Test Tool
+      run: pypy2 pytest.py rpython/tool
+
+  test-jit-x86:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-20.04]
+
+    steps:
+    - uses: actions/checkout@v4
+    - uses: actions/setup-python@v5
+      with:
+        python-version: "pypy2.7-7.3.10"
+    - run: |
+          echo "Install requirements"
+          sudo apt-get install -y libgc-dev
+          pypy2 -m pip install --user hypothesis==4.39.3
+    - name: Test Jit - X86
+      run: pypy2 pytest.py -vs rpython/jit/backend/x86
+
+  test-jit-other:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-20.04]
+
+    steps:
+    - uses: actions/checkout@v4
+    - uses: actions/setup-python@v5
+      with:
+        python-version: "pypy2.7-7.3.10"
+    - run: |
+          echo "Install requirements"
+          sudo apt-get install -y libgc-dev
+          pypy2 -m pip install --user hypothesis==4.39.3
+    - name: Test Jit - Other
+      run: pypy2 pytest.py -vs rpython/jit --ignore=rpython/jit/backend/x86
```

Comment:
This is going to be quite slow single-threaded. How long does it take to run locally?

Reply:
Here's a fully passing run where I replace pypy with cpython for the jit tests: https://github.com/olliemath/pypy-official/actions/runs/7420947044/job/20193310402

The timings are:

misc (pypy): 1 minute
**jit-x86 (cpython): 47 minutes**
jit-other (cpython): 18 minutes
memory (pypy): 10 minutes
rlib (pypy): 22 minutes
rtyper (pypy): 8 minutes
translate (pypy): 28 minutes

Under pypy those same jit tests are timing out after 6 hours (see here https://github.com/olliemath/pypy-official/actions/runs/7417566997/job/20184243634 ). Doesn't seem to happen on windows so I assume it's something specific to linux+pypy

The runners have 2 cores so could definitely speed that up

Reply:
Locally I can reproduce the timeout - at the end of some tests `rpython/tool/runsubprocess.py` occasionally hangs waiting for input from stdin